src/VideoPipe.cpp

```diff
@@ -125,30 +125,61 @@ VideoBuffer::const_shared VideoPipe::GrabFrame(uint32_t timeout)
 		return videoBuffer;
 	}
 
-	//Check if there is any new image
-	if(imgNew==0)
+	// timespec calced once outside the loop so we wait for the correct timeout duration
+	// when handling spurious wakeups
+	timespec wakeupTime;
+	if (timeout)
 	{
+		//Calculate timeout
+		calcTimout(&wakeupTime,timeout);
+	}
+
+	// Check if there is a new image, or we need to wakeup for some other reason
+	//
+	// Note: This is in a loop as is required when using pthread_cond_wait calls. You need to check for 
+	// spurious wakeups (which are uncommon) and cant just rely on signal() waking up on event
+	// See: https://pubs.opengroup.org/onlinepubs/009604599/functions/pthread_cond_timedwait.html
+	// 
+	// This means we need to check all conditions in the loop predicate that may signal the condition to wakeup including:
+	// * End() : Setting inited == false
+	// * CancelGrabFrame() : Setting cancelledGrab == true
+	// * NextFrame() : Adding a new frame to the queue
+	while(inited && queue.length()==0 && !cancelledGrab)
+	{
+		int ret = -1;
+
 		//If timeout has been specified
 		if (timeout)
 		{
-			timespec   ts;
-			//Calculate timeout
-			calcTimout(&ts,timeout);
 			//wait
-			int ret = pthread_cond_timedwait(&newPicCond,&newPicMutex,&ts);
-			if (ret && ret!=ETIMEDOUT)
-				Error("-VideoPipe cond timedwait error [%d,%d]\n",ret,errno);
-		} else {
+			ret = pthread_cond_timedwait(&newPicCond,&newPicMutex,&wakeupTime);
+			if (ret == ETIMEDOUT)
+			{
+				// On timeout exit the loop
+				break;
+			}
+		}
+		else
+		{
 			//Wait ad infinitum
-			pthread_cond_wait(&newPicCond,&newPicMutex);
+			ret = pthread_cond_wait(&newPicCond,&newPicMutex);
+		}
+		
+		if (ret)
+		{
+			Error("-VideoPipe cond wait error [%d,%d]\n",ret,errno);
 		}
 	}
 
-	//Reset new image flag
-	imgNew=0;
+	// If this was true, handle img as normal (may exist or may not) and clear this flag so we dont cancel the next one as well
+	cancelledGrab=0;
```

Comment:
In CancelGrab we reset the queue, if we want to make this explicit, i.e. a canceled grab returns no frame (which seems coherent, we should change this to

```suggestion
	if (cancelledGrab)
	{
		//Reset flag	
		cancelledGrab = 0;
		//A canceled grab returns no frame
		return videoBuffer;
	}
```